### PR TITLE
JDK-8311076: RedefineClasses doesn't check for ConstantPool overflow

### PR DESCRIPTION
**Pull request by alexmenkov**
The fix adds check that merged constant pool does not overflow u2 (two-byte unsigned).
The check is added after merging `the_class` and `scratch_class` constant pools, but before rewriting constant pool references.

testing:
 - sanity tier1;
 - all RedefineClasses/RetransformClasses tests:
   - test/jdk/java/lang/instrument
   - test/hotspot/jtreg/serviceability/jvmti/RedefineClasses
   - test/hotspot/jtreg/vmTestbase/nsk/jvmti/RedefineClasses
   - test/hotspot/jtreg/vmTestbase/nsk/jvmti/RetransformClasses

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311076](https://bugs.openjdk.org/browse/JDK-8311076): RedefineClasses doesn't check for ConstantPool overflow (**Bug** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17759/head:pull/17759` \
`$ git checkout pull/17759`

Update a local copy of the PR: \
`$ git checkout pull/17759` \
`$ git pull https://git.openjdk.org/jdk.git pull/17759/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17759`

View PR using the GUI difftool: \
`$ git pr show -t 17759`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17759.diff">https://git.openjdk.org/jdk/pull/17759.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17759#issuecomment-1932880693)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp [hunk 1]
```diff
@@ -1823,6 +1823,11 @@ jvmtiError VM_RedefineClasses::merge_cp_and_rewrite(
     return JVMTI_ERROR_INTERNAL;
   }
 
+  // ensure merged constant pool size does not overflow u2
+  if (merge_cp_length > 0xFFFF) {
+    return JVMTI_ERROR_INTERNAL;
```
Comment (lmesnik) [inline src/hotspot/share/prims/jvmtiRedefineClasses.cpp:1828]: Doesn't it make a sense to add some logging there so user could easier understand the problem?

Comment (coleenp) [reply src/hotspot/share/prims/jvmtiRedefineClasses.cpp:1828]: Seems like a good idea, you can use log_warning(redefine, class, constantpool)("...")

**Review by coleenp — APPROVED** [review]
Thank you for fixing this.  Looks good.

**Review by plummercj — COMMENTED** [review]
It looks like we don't have a reproducer for this issue. Please add the appropriate noreg label.

**Review by lmesnik — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp (hunk 1 above)
Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiRedefineClasses.cpp:1829]: Done.

**Review by coleenp — APPROVED** [review]
Yes, this should definitely be tagged noreg-hard.

**Pull request closed without merging**